rfcs/0146-meta-categories.md

```diff
@@ -0,0 +1,204 @@
+
+1. It is bounded by the constraints imposed by the filesystem.
+
+   - Restrictions on filenames, subdirectory tree depth, permissions, inodes,
+     quotas, and many other things.
+     - Some of these restrictions are not well documented and are found simply
+       by "bumping" on them.
+     - The restrictions can vary on an implementation basis.
+       - Some filesystems have more restrictions or less features than others,
+         forcing an uncomfortable lowest common denominator.
+       - Some operating systems can impose additional constraints over otherwise
+         full-featured filesystems because of backwards compatibility (8 dot
+         3,anyone?).
+
+2. It requires a local checkout of the tree.
+
+   Certainly this checkout can be "cached" using some form of `find . >
+   /tmp/pkgs-listing.txt`, or more sophisticated solutions like `locate +
+   updatedb`. Nonetheless such solutions still require access to a fresh,
+   updated copy of the Nixpkgs tree.
+
+3. The creation of a new category - and more generally the manipulation of
+   categories - requires an unpleaseant task of renaming and eventually patching
+   many seemingly unrelated files.
+
+   - Moving files around Nixpkgs codebase requires updating their forward and
+     backward references.
+     - Especially in some auxiliary tools like editor plugins, testing suites,
+       autoupdate scripts and so on.
+   - Rewriting `all-packages.nix` can be error-prone (even using Metapad) and it
+     can generate huge, noisy patches.
+
+4. There is no convenient way to use multivalued categorization.
+
+   A piece of software can fulfill many categories; e.g. 
+   - an educational game
+   - a console emulator (vs. a PC emulator)
+   - and a special-purpose programming language (say, a smart-contracts one).
+
+   The current one-size-fits-all restriction is artificial, imposes unreasonable
+   limitations and results in incomplete and confusing information.
+
+   - No, symlinks or hardlinks are not convenient for this purpose; not all
+     environments support them (falling on the "less features than others"
+     problem expressed before) and they convey nothing besides confusion - just
+     think about writing the corresponding entry in `all-packages.nix`.
+
+5. It puts over the (possibly human) package writer the mental load of where to
+   put the files on the filesystem hierarchy, deviating them from the job of
+   really writing them.
+
+   - Or just taking the shortest path and throw it on a folder under `misc`.
+
+6. It "locks" the filesystem, preventing its usage for other, more sensible
+   purposes.
+
+7. The most important: the categorization is not discoverable via Nix language
+   infrastructure.
+
+   Indeed there is no higher level way to query about such categories besides
+   the one described in the bullet 2 above.
+
+In light of such a bunch of problems, this RFC proposes a novel alternative to
+the above mess: new `meta` attributes.
+
```

Comment:
Would be nice to have some motivation why we want categories in the first place, since it will inform the discussion/implementation.

If the primary motivation was just dealing with filesystem constraints, categories will not be relevant with RFC #140, and nuking them might be a good idea.

Maybe we want to make package discovery easier. But is the effort to implement categories worth it? How often do people look for a software in a specific category? I typically look elsewhere first e.g. at [topic-specific curated lists](https://github.com/sindresorhus/awesome) and then just search Nixpkgs for a specific software project I liked.

Reply:
Well, categories are very traditional, many other package databases include them. As an example, in NetBSD pkgsrc it is even mandatory:

https://github.com/NetBSD/pkgsrc/blob/trunk/mk/bsd.pkg.mk#L130

And this specific directive is very old, older than NixOS itself - here is a similar code from 2022-12-31:

https://github.com/NetBSD/pkgsrc/blob/7751a84cdeb294f49bb9143d4e502a0354d75006/mk/bsd.pkg.mk#L845

Therefore, the weight of tradition is very important to keep it.

(Good! I will include this!)

----------

Personally I don't have a great idea of use case.

Maybe I am just curious and willing to test a different software from a similar category - like terminal emulators or window managers.

Maybe I want to do a curated list of a certain type of software - games and emulators I want to put on a custom ISO.

Maybe I want to make a comparison about softwares on a same category for a research project - then I can just throw a term like "debugger" or "programming language" and collect the results.

Reply:
I can see a few use cases:

1. It makes searching for some stuff easier. Currently, I'm going to search.nixos.org , I enter "sqlite" and then search for a package I need. But I could just enter "sqlite" and pick category "python" to see all of the Python libraries related to sqlite. Easier than picking a query that would search for both, more accurate than prefixing the query with `python311Packages` (also one shouldn't have to remember that namespace name)
2. It makes indexing easier. Currently we only have search.nixos.org which uses some arcane sorcery to do that. If we have categories/tags, indexing is trivial with any language and any database
3. Those two factors make offline package searching sane

Reply:
The "feature" already exists. I am not advocating for a resurrection of anything already dead, but for a better interface for something that already exists.

And, hey, are you tired of traditions? Try to reopen RFC #122, your "it is like resurrecting FHS" argument will be very convincing! XD

Reply:
I don't understand the remark on RFC112, letting this aside

You say:

> Personally I don't have a great idea of use case.

And I think that @jtojnar remark on having usecases before doing busywork is probably a good idea.

Now, it seems like people are coming up with usecases and I feel the RFC would become more interesting by writing them down and weighting them down, my point is only that tradition is not a good enough reason to include a feature.

Reply:
> I don't understand the remark on RFC112, letting this aside
> 
> You say:
> 
> > Personally I don't have a great idea of use case.
> 

I have ideas, I show them. I just said they aren't so great.

> And I think that @jtojnar remark on having usecases before doing busywork is probably a good idea.
>

Maybe. 

But I don't want to divert the purpose of this RFC. I will not do a huge, full-fledged, hardcore defense of categorization, given that everyone does categorizations even in the current Nixpkgs codebase.

> my point is only that tradition is not a good enough reason to include a feature.

As I said, I am not including an inexistent thing, but enhancing something already existent.

Further, I am not treating the tradition argument as an all-or-nothing argument.

And if you don't like the word "tradition", change it to "standard".